pkg/util/kube/utils.go

```diff
@@ -35,18 +38,53 @@ func NamespaceAndName(objMeta metav1.Object) string {
 	return fmt.Sprintf("%s/%s", objMeta.GetNamespace(), objMeta.GetName())
 }
 
-// EnsureNamespaceExists attempts to create the provided Kubernetes namespace. It returns two values:
-// a bool indicating whether or not the namespace was created, and an error if the create failed
+// EnsureNamespaceExistsAndIsReady attempts to create the provided Kubernetes namespace. It returns two values:
+// a bool indicating whether or not the namespace is ready, and an error if the create failed
 // for a reason other than that the namespace already exists. Note that in the case where the
-// namespace already exists, this function will return (false, nil).
-func EnsureNamespaceExists(namespace *corev1api.Namespace, client corev1client.NamespaceInterface) (bool, error) {
-	if _, err := client.Create(namespace); err == nil {
+// namespace already exists and is not ready, this function will return (false, nil).
+// If the namespace exists and is marked for deletion, this function will wait up to the timeout for it to fully delete.
+func EnsureNamespaceExistsAndIsReady(namespace *corev1api.Namespace, client corev1client.NamespaceInterface, timeout time.Duration) (bool, error) {
+	var ready bool
+	err := wait.PollImmediate(time.Second, timeout, func() (bool, error) {
+		clusterNS, err := client.Get(namespace.Name, metav1.GetOptions{})
+
+		if apierrors.IsNotFound(err) {
+			// Namespace isn't in cluster, we're good to create.
+			return true, nil
+		}
+
+		if err != nil {
+			// Return the err and exit the loop.
+			return true, err
+		}
+
+		if clusterNS != nil && (clusterNS.GetDeletionTimestamp() != nil || clusterNS.Status.Phase == corev1api.NamespaceTerminating) {
+			// Marked for deletion, keep waiting
+			return false, nil
+		}
+
+		// clusterNS found, is not nil, and not marked for deletion, therefore we shouldn't create it.
+		ready = true
 		return true, nil
-	} else if apierrors.IsAlreadyExists(err) {
-		return false, nil
-	} else {
+	})
+
+	// err will be set if we timed out or encountered issues retrieving the namespace,
+	if err != nil {
+		return false, errors.Wrapf(err, "error getting namespace %s", namespace.Name)
+	}
+
+	// In the case the namespace already exists and isn't marked for deletion, assume it's ready for use.
+	if ready {
+		return true, nil
+	}
+
+	if _, err := client.Create(namespace); err != nil {
+		// if we get any error here, something else created the NS after all of our polling
```

Comment:
Do you want to check for an already-exists error?

Reply:
So I think we can, but does that mean we should check the phase and deletion timestamp again? Do we care if something else made the namespace before we got to?

Reply:
No, I don't think so. The reason I suggested it was to handle
1. we decide we need to create the namespace (it doesn't exist, or it did and we waited for it to be deleted)
1. something else creates the namespace
1. we try to create the namespace, and it fails because it already exists

If this is the sequence, I don't think we should return an error.

If this is the sequence, *and* the namespace somehow is terminating, I think it's ok to return an error. I don't expect this to happen, and I don't want to get into an infinite loop 😄